ble_app.h

```diff
@@ -0,0 +1,427 @@
+     * This is called by Gap to notify the application we disconnected
+     */
+    void onDisconnectionComplete(const ble::DisconnectionCompleteEvent &event) override
+    {
+        if (_connected) {
+            _connected = false;
+            printf("Disconnected.\r\n");
+            _event_queue.call([this]() { start_activity(); });
+        }
+    }
+
+    /** Restarts main activity */
+    void onAdvertisingEnd(const ble::AdvertisingEndEvent &event)
+    {
+        _event_queue.call([this]() { start_activity(); });
+    }
+
+    /**
+     * Start advertising or scanning. Triggered by init or disconnection.
+     */
+    virtual void start_activity()
+    {
+        if (!_ble.hasInitialized()) {
+            return;
+        }
+
+        if (_advertising_name) {
+            start_advertising();
+        } else {
+            _ble.gap().stopAdvertising(_adv_handle);
+        }
+
+        if (_target_name) {
+            start_scanning();
+        } else {
+            _ble.gap().stopScan();
+        }
+    }
+
+    /**
+     * Start the advertising process; it ends when a device connects.
+     */
+    void start_advertising()
+    {
+        ble_error_t error;
+
+        if (!_advertising_name || _ble.gap().isAdvertisingActive(_adv_handle)) {
+            /* we're already advertising */
+            return;
+        }
+
+        ble::AdvertisingParameters adv_params(
+            ble::advertising_type_t::CONNECTABLE_UNDIRECTED,
+            ble::adv_interval_t(ble::millisecond_t(40))
+        );
+
+        error = _ble.gap().setAdvertisingParameters(_adv_handle, adv_params);
+
+        if (error) {
+            printf("_ble.gap().setAdvertisingParameters() failed\r\n");
+            return;
+        }
+
+        uint8_t adv_buffer[MAX_ADVERTISING_PAYLOAD_SIZE];
+        ble::AdvertisingDataBuilder adv_data_builder(adv_buffer);
+        
+        adv_data_builder.clear();
+        adv_data_builder.setFlags();
+        error = adv_data_builder.setName(_advertising_name);
+
+        if (error) {
+            print_error(error, "AdvertisingDataBuilder::setName() failed (name too long?)\r\n");
+            return;
+        }
+
+        /* Set payload for the set */
+        error = _ble.gap().setAdvertisingPayload(
+            _adv_handle, adv_data_builder.getAdvertisingData()
+        );
+
+        if (error) {
+            print_error(error, "Gap::setAdvertisingPayload() failed\r\n");
+            return;
+        }
+
+        error = _ble.gap().startAdvertising(_adv_handle, ble::adv_duration_t(ble::second_t(10)));
+
+        if (error) {
+            print_error(error, "Gap::startAdvertising() failed\r\n");
+            return;
+        }
+
+        printf("Advertising as \"%s\"\r\n", _advertising_name);
+    }
+
+    /** scan for GattServer */
+    void start_scanning()
+    {
+        if (_is_scanning || _connected || !_target_name) {
+            /* already connected or scan not needed */
```

Comment:
Perhaps you could leave a trace?

Reply:
not really an error, just normal flow as the activity is started from many places which don't know what the current state is.